### ISSUE TITLE
Broaden the scope?

### ISSUE THREAD

**Issue by georges** [hessu/aprs-specs#5]
First, thanks for publishing these excellent documents. I've implemented the specification for BLE and TCP TNC discovery in an iOS Winlink client. It is magical!

A couple of thoughts:

1.  The KISS TNC discovery spec described here are broader than APRS. Perhaps rename this repo to reflect the general purpose nature?
2.  Maybe the DNS-SD mechanism could be expanded to other type of service discovery. Two popular come to mind:

- VARA modems. Client exposes a command port, usually 8300 and a data port 8301. Since there are two distinct clients for HF and FM, and they can run simultaneously on different ports, perhaps could be published under `_vara-hf._tcp` and `_vara-fm._tcp` respectively.

- rigctl/rigctld. Usually exposed on 4532 port. Maybe `_rigctl._tcp`

Thoughts?

**Comment by hessu** [comment]
Thanks! I'm very happy to see additional implementations of these protocols. When actually using them to connect, they surely feel like sufficiently advanced technology - a sudden jump from the 1990's to 2010 or so! :)

1. Yes, sounds like a good idea. Suggestions for a name? hamradio-specs ?
2. Yes, I was thinking from the beginning that the DNS-SD approach should be used for all modem interfaces like this, as well as logging programs, SDRs and other rigs on the network, antenna rotators and so on. Once a few popular applications have started used it, it should catch up pretty quickly, as it is not that hard to implement, and makes the setup phase so much easier.

But I don't know if we should start allocating identifiers here for everyone, except for the common protocols (some protocols shared by multiple logging programs come to mind).

The DNS-SD spec probably has a documented method to communicate both a data port and a command port in a single announcement. It should be used for the VARA modem. A single node can make multiple announcements using the same protocol (say, _vara-modem._tcp) for multiple services (HF and FM), which makes sense if they are using the same control protocol for the client and a client could choose to connect to either the HF or FM modem, or both - they just need to do a single search for _vara-modem._tcp instead of two different searches.

**Comment by georges** [comment]
1. `hamradio-zeroconfig-specs` ? I know it's a mouthful but in the same time it narrows the scope of what these specifications are trying to accomplish.
2. I am not too familiar with DNS-SD capabilities. I agree that if we can keep a high level service descriptor like vara and tuck all the endpoint underneath that would be ideal. I have an interest in that one, so I may take a stab at modeling the announcement based on what you described and see how that works.

WRT allocating identifiers, it's a lot easier if a list is kept centrally, so perhaps it's ok to take a tack at defining the most popular one and invite other for comments? The nice thing is that if client side supported is added first, users can use the cmd line tools to announce the service on the host and still benefit until the application themselves add native announcement support.

You may enjoy seeing DNS-SD and BLE discovery in action in the short video I recorded yesterday. Don't blink it goes by fast :)
[https://www.youtube.com/watch?v=BE3kCXP4Brc](https://www.youtube.com/watch?v=BE3kCXP4Brc)

**Comment by hessu** [comment]
@georges Hi, I see that you implemented _varafm-modem._tcp in Varanny for the vara modem service. Nice!

I see you chose to announce _varafm-modem._tcp and _varahf-modem._tcp separately. Do the HF and FM modems speak a different protocol to the TCP client, or is there some other reason to have a different service? I would have guessed they'd have the same TCP protocol and it'd be simpler to have the client do a single search for _vara-modem._tcp which would return all FM, HF, and other future variants without adding separate queries for them in the clients. The textual description and additional TXT fields could then describe the details, and make it future-proof - support future variants without changing clients?

I suppose a doc should be started for documenting the allocation of these identifiers, and best practices. Here's my initial stab at it: https://github.com/hessu/aprs-specs/blob/b1ff96c353ddb8e33deeef1e278781790f8fcc23/hamradio-zeroconfig.md

**Comment by georges** [comment]
The protocol is similar in the sense that both fm and hf accept commands on one port and data on another. However the command set is a bit different as some commands only exist for one mode or the other.

The key design decision to keep them separate was as follow:

1. Simplicity. HF and FM are different mode each in a separate software application. If a client is looking for a type of connection, it's simpler to look for one or the other.
2. Allow for different names to be advertised that could potentially match the radios, e.g. IC-7300 and TH-D74 for example.
3. Use the port advertised in dns-sd directly to represent the main cmd port of the modem (each modem can run on a different port number). 
4. Prevent from having to add a lot of custom parsing of the TXT fields to fish out mode specific information as most parameter would have to be duplicated and name spaced either HF or FM. 

Basically, the idea was that even tho they are both called VARA, they are fundamentally different modems.

**Comment by hessu** [comment]
I suppose points 2 and 3 or not really valid, as a single node / advertiser can advertise multiple services at the same time, with the same name _var-modem._tcp, with different user-friendly names and ports, that are on the same IP.

If the client protocols are truly different and a client needs to implement them differently, then a different service name might be in place.

I guess an analogy is a printer, there are lot of different printers out there, with quite noticeable physical differences, but as long as they speak the same protocol to the client they are advertised on the same name. New printer models will appear to old clients without the clients necessarily needing updates, these days. The DNS-SD advertisement contains enough metadata for most old clients to adjust to the new printer model that they didn't know about beforehand.

With the TCP KISS, you can use 300, 1200, 9600 bit/s modems, where the actual modulator might be using AFSK or GMSK or G3RUH modulation; or indeed LoRa these days, the client does not really need to know as long as we will be doing KISS to transfer the payload. The speed and modulation can be announced in TXT fields as arbitrary strings so that they can be shown to the user. Someone implemented a LoRa modem for it and the aprs.fi app works with it, without me lifting a finger.

**Comment by georges** [comment]
You're making a good point. I think I mistakenly assumed that a device would only be able to advertise on a given service once. Although I'm not sure why because, I do advertise multiple entries for FM radios. I vaguely remember starting with just one name and tried to cram everything in the TXT field and backed away. 

If I understand correctly, we could conceive of a scheme where

VARA FM
`Radio A _vara-modem._tcp port 8300 TXT="type=fm, ..."`
`Radio B _vara-modem._tcp port 8300 TXT="type=fm, ..."`

VARA HF
`Radio C _vara-modem._tcp port 8400 TXT="type=hf, ..."`
`Radio D _vara-modem._tcp port 8400 TXT="type=hf, ..."`

and then the other TXT attributes would be common, so my point 4 is mute.

I think the only drawback is that for a client to filter fm/hf when scanning for devices it needs to resolve to get the TXT field, which would add some delay.

I'm tempted to explore this further and refactor the code accordingly. Will experiment when I get some time.

**Comment by hessu** [comment]
I suspect the TXT field will come to the client in the same packet with the other stuff and does not add a delay, unless it's very big. But I'm not absolutely sure.

This sort of approach would hopefully make addition of future "VARA Satellite" or "VARA EME" modems transparent to the clients, which I think is the main beef in the whole exercise - new features and modems just magically working in existing clients without modification.

**Comment by georges** [comment]
Ok I refactored `varanny` to now use `_vara-modem._tcp` as a service name and specify the modem type in the TXT. Thanks for pointing this out, the implementation feels a lot cleaner now.

**Comment by hessu** [comment]
Cool, I'm happy to hear you feel that way.

I'll adjust the allocation document accordingly?
